### PR TITLE
 Page-Hinkley test stream aggregate implemented. 

### PR DESCRIPTION
**Pull request by klemenkenda**

### CODE REVIEW

File: src/nodejs/qm/qm_nodejs_streamaggr.h [hunk 1]
```diff
@@ -1817,6 +1818,100 @@
...
+*   has been detected in this step, otherise to 0. `'drift`' is always set to 0 in the next step. `'driftOffset'` 
+*   monitors the offset (number of onStep calls) since last concept drift was detected.
+* <br>2. {@link module:qm.StreamAggr#getParams} returns a parameter object.
+* <br>3. {@link module:qm.StreamAggr#setParams} used for changing Page-Hinkley test parameters
+*
+* @property {string} [name] - The given name of the stream aggregator (autogenerated by default).
+* @property {string} type - The type for the stream aggregator. <b>Important:</b> It must be equal to `'pagehinkley'`.
+* @property {string} store - The name of the store consistent with the records that it processes.
+* @property {string} inAggr - The name of the stream aggregate used for input (i.e. {@link module:qm.StreamAggrTimeSeriesTick}).
+* @property {number} maxInstances - Minimal number of instances needed for initialization of the aggregator (when can first concept drift be initialized?).
```
Comment (rupnikj) [inline src/nodejs/qm/qm_nodejs_streamaggr.h]: rename to `minInstances`?

File: src/qminer/qminer_aggr.cpp [hunk 2]
```diff
@@ -2781,5 +2781,126 @@ PJsonVal THistogramAD::SaveJson(const int& Limit) const {
...
 
+
+//////////////////////////////////////////////////////////////////////////////////
+/// Page-Hinkley test for concept drift detection aggregate
+
+void TPageHinkley::OnStep(const TWPt<TStreamAggr>& CallerAggr) {	
+	TScopeStopWatch StopWatch(ExeTm);
+	
+	if (Drift > 0) Reset();
+	TFlt X = InAggrVal->GetFlt();
```
Comment (rupnikj) [inline src/qminer/qminer_aggr.cpp]: double X
TFlt is to be used in class members

File: src/qminer/qminer_aggr.cpp [hunk 3]
```diff
@@ -2781,5 +2781,126 @@ PJsonVal THistogramAD::SaveJson(const int& Limit) const {
...
+	XMean = XMean + (X - XMean) / (1.0 * SampleCount);
+	Sum = Alpha * Sum + (X - XMean - Delta);
+	
+	SampleCount++;
+	
+	/// Only update drift offset, if it has already occured
+	if (DriftOffset >= 0) DriftOffset++;
```
Comment (rupnikj) [inline src/qminer/qminer_aggr.cpp]: { }

File: src/qminer/qminer_aggr.cpp [hunk 4]
```diff
@@ -2781,5 +2781,126 @@ PJsonVal THistogramAD::SaveJson(const int& Limit) const {
...
+	Alpha.Save(SOut);
+	XMean.Save(SOut);
+	SampleCount.Save(SOut);
+	Sum.Save(SOut);
+	Drift.Save(SOut);
+	DriftOffset.Save(SOut);
+}
+
+void TPageHinkley::Reset() {
+	MinInstances = 30;
```
Comment (rupnikj) [inline src/qminer/qminer_aggr.cpp]: Reset should not change parameters to default - it should just change the state to initial state

File: src/qminer/qminer_aggr.cpp [hunk 5]
```diff
@@ -2781,5 +2781,126 @@ PJsonVal THistogramAD::SaveJson(const int& Limit) const {
...
+	Delta = (double)ParamVal->GetObjNum("delta", 0.005);
+	Lambda = (double)ParamVal->GetObjNum("lambda", 50.0);
+	Alpha = (double)ParamVal->GetObjNum("alpha", 0.9999);
+}
+
+void TPageHinkley::LoadState(TSIn& SIn) {	
+	MinInstances.Load(SIn);
+	Delta.Load(SIn);
+	Lambda.Load(SIn);
+	Alpha.Load(SIn);
```
Comment (rupnikj) [inline src/qminer/qminer_aggr.cpp]: LoadState and SaveState should only save state not algorithm parameters
Example: TAggrResampler 
https://github.com/qminer/qminer/blob/ef660b04651959a85d2ed870e7860f4e560ea70f/src/glib/mine/signalproc.cpp#L938

File: src/qminer/qminer_aggr.cpp [hunk 6]
```diff
@@ -2781,5 +2781,126 @@ PJsonVal THistogramAD::SaveJson(const int& Limit) const {
...
+	Alpha = 0.9999;
+	XMean = 0.0;
+	SampleCount = 1;
+	Sum = 0.0;
+	Drift = 0;
+	DriftOffset = 0;
+}
+
+int TPageHinkley::GetInt() const {
+	return GetNmInt("drift");
```
Comment (rupnikj) [inline src/qminer/qminer_aggr.cpp]: tabs

File: src/qminer/qminer_aggr.h [hunk 7]
```diff
@@ -2199,6 +2199,77 @@ class THistogramAD : public TStreamAggr, public TStreamAggrOut::INmInt, public T
...
 
+//////////////////////////////////////////////////////////////////////////////////
+/// Page-Hinkley test for concept drift detection aggregate
+///
+/// This change detection method works by computing the observed values and their mean up to the current
+/// moment. Page Hinkley won't output warning zone warnings, only change detections. The method works by
+/// means of the Page Hinkley test.In general lines it will detect a concept drift if the observed mean
+/// at some instant is greater then a threshold value lambda.
+class TPageHinkley : public TStreamAggr, 
+					 public TStreamAggrOut::INmInt,
```
Comment (rupnikj) [inline src/qminer/qminer_aggr.h]: tabs

File: src/qminer/qminer_core.cpp [hunk 8]
```diff
@@ -6612,6 +6612,7 @@ void TStreamAggr::Init() {
     Register<TStreamAggrs::TWinBufSpVecSum>();
     Register<TStreamAggrs::TRecSwitchAggr>();
     Register<TStreamAggrs::THistogramAD>();
+	Register<TStreamAggrs::TPageHinkley>();
```
Comment (rupnikj) [inline src/qminer/qminer_core.cpp]: tabz

File: test/nodejs/streamaggr2.js [hunk 9]
```diff
@@ -1289,20 +1289,204 @@ describe('Simple linear regression test', function () {
...
         var res2 = linReg.saveJson();
         assert.equal(JSON.stringify(res), JSON.stringify(res2));
-            
+
         })
     });
 });
+
+describe("test concept drift methods", function() {
+    let qm = require('qminer');
```
Comment (rupnikj) [inline test/nodejs/streamaggr2.js]: this should be on top of the file

File: test/nodejs/streamaggr2.js [hunk 10]
```diff
@@ -1289,20 +1289,204 @@ describe('Simple linear regression test', function () {
...
+            // creating start time
+            let time = new Date();
+            let timeStr = time.toISOString();
+            let changes = 0;
+            time.setSeconds(time.getSeconds() - 1000);
+
+            for (let i = 0; i < 1000; i++) {
+                // add one second to the timestamp and create an ISO string
+                time.setSeconds(time.getSeconds() + 1);
+                let timeStr = time.toISOString();
```
Comment (rupnikj) [inline test/nodejs/streamaggr2.js]: remove let or line 1379

File: test/nodejs/streamaggr2.js [hunk 11]
```diff
@@ -1289,20 +1289,204 @@ describe('Simple linear regression test', function () {
...
+                // create value
+                let value = Math.random() * 1;
+                // adding values to the signal store
+                store.push({ Time: timeStr, Value: value });
+                // counting changes
+                if (pht.saveJson().drift == 1) {
+                    changes++;
+                }
+            }
+            // checking if drift has been correctly detected
```
Comment (rupnikj) [inline test/nodejs/streamaggr2.js]: checking that drift has not been detected

File: test/nodejs/streamaggr2.js [hunk 12]
```diff
@@ -1289,20 +1289,204 @@ describe('Simple linear regression test', function () {
...
+            assert.equal(changes, 1);
+            driftOffset = pht.saveJson().driftOffset;
+            assert.equal((driftOffset < 1000) && (driftOffset > 900), true);
+        })
+
+        it('should save and then load the aggregate', function() {
```
Comment (rupnikj) [inline test/nodejs/streamaggr2.js]: add two tests:
- save/load uninitialized aggregate
- save/load initialized aggregate with data

Comment (rupnikj) [reply test/nodejs/streamaggr2.js]: Example: https://github.com/rupnikj/qminer/blob/master/test/nodejs/streamaggr.js#L6080

File: test/nodejs/streamaggr2.js [hunk 13]
```diff
@@ -1289,20 +1289,204 @@ describe('Simple linear regression test', function () {
...
+            // creating start time
+            let time = new Date();
+            let timeStr = time.toISOString();
+            let changes = 0;
+
+            // simulating concept drift at element 1000 in a time series
+            for (let i = 0; i < 2000; i++) {
+                // add one second to the timestamp and create an ISO string
+                time.setSeconds(time.getSeconds() + 1);
+                let timeStr = time.toISOString();
```
Comment (rupnikj) [inline test/nodejs/streamaggr2.js]: let

File: src/nodejs/qm/qm_nodejs_streamaggr.h [hunk 14]
```diff
@@ -1817,6 +1818,100 @@
...
+* // creating start time
+* let time = new Date();
+* let timeStr = time.toISOString();
+* let changes = 0;
+*
+* // simulating concept drift at element 1000 in a time series
+* for (let i = 0; i < 2000; i++) {
+*     // add one second to the timestamp and create an ISO string
+*     time.setSeconds(time.getSeconds() + 1);
+*     let timeStr = time.toISOString();
```
[the remark this answers is not in the document]
Comment (klemenkenda) [reply src/nodejs/qm/qm_nodejs_streamaggr.h]: Done.

File: src/nodejs/qm/qm_nodejs_streamaggr.h (hunk 1 above)
Comment (klemenkenda) [reply src/nodejs/qm/qm_nodejs_streamaggr.h]: Done.

File: src/qminer/qminer_aggr.cpp [hunk 15]
```diff
@@ -2781,5 +2781,126 @@ PJsonVal THistogramAD::SaveJson(const int& Limit) const {
...
+
+//////////////////////////////////////////////////////////////////////////////////
+/// Page-Hinkley test for concept drift detection aggregate
+
+void TPageHinkley::OnStep(const TWPt<TStreamAggr>& CallerAggr) {	
+	TScopeStopWatch StopWatch(ExeTm);
+	
+	if (Drift > 0) Reset();
+	TFlt X = InAggrVal->GetFlt();
+	XMean = XMean + (X - XMean) / (1.0 * SampleCount);
```
Comment (rupnikj) [inline src/qminer/qminer_aggr.cpp]: 1.0 -> (double)

File: test/nodejs/streamaggr2.js (hunk 12 above)
Comment (klemenkenda) [reply test/nodejs/streamaggr2.js]: Made 2 tests. Is this OK?

**Pull request merged**